File: backend/app/schemas/store.py
"""Store API schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.store_item import (
    StoreItemCurrency,
    StoreItemType,
)


# ============================================================
# Store Category
# ============================================================

class StoreCategoryCreate(BaseModel):
    name_ar: str = Field(
        min_length=1,
        max_length=100,
    )
    name_en: str = Field(
        min_length=1,
        max_length=100,
    )
    description_ar: str | None = None
    description_en: str | None = None
    icon_url: str | None = Field(
        default=None,
        max_length=500,
    )
    image_url: str | None = Field(
        default=None,
        max_length=500,
    )
    sort_order: int = 0
    is_active: bool = True


class StoreCategoryUpdate(BaseModel):
    name_ar: str | None = Field(
        default=None,
        min_length=1,
        max_length=100,
    )
    name_en: str | None = Field(
        default=None,
        min_length=1,
        max_length=100,
    )
    description_ar: str | None = None
    description_en: str | None = None
    icon_url: str | None = Field(
        default=None,
        max_length=500,
    )
    image_url: str | None = Field(
        default=None,
        max_length=500,
    )
    sort_order: int | None = None
    is_active: bool | None = None


class StoreCategoryRead(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: UUID
    name_ar: str
    name_en: str
    description_ar: str | None
    description_en: str | None
    icon_url: str | None
    image_url: str | None
    sort_order: int
    is_active: bool
    created_at: datetime
    updated_at: datetime


# ============================================================
# Store Item
# ============================================================

class StoreItemCreate(BaseModel):
    category_id: UUID

    name_ar: str = Field(
        min_length=1,
        max_length=150,
    )
    name_en: str = Field(
        min_length=1,
        max_length=150,
    )

    description_ar: str | None = None
    description_en: str | None = None

    item_type: StoreItemType
    currency: StoreItemCurrency

    price: int = Field(
        ge=0,
    )

    image_url: str | None = Field(
        default=None,
        max_length=500,
    )

    asset_id: str | None = Field(
        default=None,
        max_length=150,
    )

    duration_days: int | None = Field(
        default=None,
        gt=0,
    )

    stock: int | None = Field(
        default=None,
        ge=0,
    )

    sort_order: int = 0
    is_active: bool = True


class StoreItemUpdate(BaseModel):
    category_id: UUID | None = None

    name_ar: str | None = Field(
        default=None,
        min_length=1,
        max_length=150,
    )
    name_en: str | None = Field(
        default=None,
        min_length=1,
        max_length=150,
    )

    description_ar: str | None = None
    description_en: str | None = None

    item_type: StoreItemType | None = None
    currency: StoreItemCurrency | None = None

    price: int | None = Field(
        default=None,
        ge=0,
    )

    image_url: str | None = Field(
        default=None,
        max_length=500,
    )

    asset_id: str | None = Field(
        default=None,
        max_length=150,
    )

    duration_days: int | None = Field(
        default=None,
        gt=0,
    )

    stock: int | None = Field(
        default=None,
        ge=0,
    )

    sort_order: int | None = None
    is_active: bool | None = None


class StoreItemRead(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: UUID
    category_id: UUID

    name_ar: str
    name_en: str

    description_ar: str | None
    description_en: str | None

    item_type: StoreItemType
    currency: StoreItemCurrency

    price: int

    image_url: str | None
    asset_id: str | None
    duration_days: int | None
    stock: int | None

    sort_order: int
    is_active: bool

    created_at: datetime
    updated_at: datetime


# ============================================================
# Store Banner
# ============================================================

class StoreBannerCreate(BaseModel):
    title_ar: str | None = Field(
        default=None,
        max_length=200,
    )
    title_en: str | None = Field(
        default=None,
        max_length=200,
    )

    image_url: str = Field(
        min_length=1,
        max_length=500,
    )

    target_url: str | None = Field(
        default=None,
        max_length=500,
    )

    sort_order: int = 0
    is_active: bool = True

    starts_at: datetime | None = None
    ends_at: datetime | None = None


class StoreBannerUpdate(BaseModel):
    title_ar: str | None = Field(
        default=None,
        max_length=200,
    )
    title_en: str | None = Field(
        default=None,
        max_length=200,
    )

    image_url: str | None = Field(
        default=None,
        min_length=1,
        max_length=500,
    )

    target_url: str | None = Field(
        default=None,
        max_length=500,
    )

    sort_order: int | None = None
    is_active: bool | None = None

    starts_at: datetime | None = None
    ends_at: datetime | None = None


class StoreBannerRead(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: UUID

    title_ar: str | None
    title_en: str | None

    image_url: str
    target_url: str | None

    sort_order: int
    is_active: bool

    starts_at: datetime | None
    ends_at: datetime | None

    created_at: datetime
    updated_at: datetime

File: admin/src/types/store.ts
/**
 * Admin Store module.
 * Mirrors `backend/app/schemas/store.py` and the enums in
 * `backend/app/models/store_item.py`.
 */

export const STORE_ITEM_TYPES = ['AVATAR', 'FRAME'] as const
export type StoreItemType = (typeof STORE_ITEM_TYPES)[number]

export const STORE_CURRENCIES = ['POINTS', 'COINS'] as const
export type StoreCurrency = (typeof STORE_CURRENCIES)[number]

// ============================================================
// Store Categories
// ============================================================

export type StoreCategoryRead = {
  id: string
  name_ar: string
  name_en: string
  description_ar: string | null
  description_en: string | null
  icon_url: string | null
  image_url: string | null
  sort_order: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export type StoreCategoryCreatePayload = {
  name_ar: string
  name_en: string
  description_ar?: string | null
  description_en?: string | null
  icon_url?: string | null
  image_url?: string | null
  sort_order?: number
  is_active?: boolean
}

export type StoreCategoryUpdatePayload = {
  name_ar?: string | null
  name_en?: string | null
  description_ar?: string | null
  description_en?: string | null
  icon_url?: string | null
  image_url?: string | null
  sort_order?: number | null
  is_active?: boolean | null
}

// ============================================================
// Store Items
// ============================================================

export type StoreItemRead = {
  id: string
  category_id: string
  name_ar: string
  name_en: string
  description_ar: string | null
  description_en: string | null
  item_type: StoreItemType
  currency: StoreCurrency
  price: number
  image_url: string | null
  asset_id: string | null
  duration_days: number | null
  available_from: string | null
  available_until: string | null
  ownership_duration_days: number | null
  stock: number | null
  sort_order: number
  is_active: boolean
  created_at: string
  updated_at: string
}

export type StoreItemCreatePayload = {
  category_id: string
  name_ar: string
  name_en: string
  description_ar?: string | null
  description_en?: string | null
  item_type: StoreItemType
  currency: StoreCurrency
  price: number
  image_url?: string | null
  asset_id?: string | null
  duration_days?: number | null
  available_from?: string | null
  available_until?: string | null
  ownership_duration_days?: number | null
  stock?: number | null
  sort_order?: number
  is_active?: boolean
}

export type StoreItemUpdatePayload = {
  category_id?: string | null
  name_ar?: string | null
  name_en?: string | null
  description_ar?: string | null
  description_en?: string | null
  item_type?: StoreItemType | null
  currency?: StoreCurrency | null
  price?: number | null
  image_url?: string | null
  asset_id?: string | null
  duration_days?: number | null
  available_from?: string | null
  available_until?: string | null
  ownership_duration_days?: number | null
  stock?: number | null
  sort_order?: number | null
  is_active?: boolean | null
}

// ============================================================
// Store Banners
// ============================================================

export type StoreBannerRead = {
  id: string
  title_ar: string | null
  title_en: string | null
  image_url: string
  target_url: string | null
  sort_order: number
  is_active: boolean
  starts_at: string | null
  ends_at: string | null
  created_at: string
  updated_at: string
}

export type StoreBannerCreatePayload = {
  title_ar?: string | null
  title_en?: string | null
  image_url: string
  target_url?: string | null
  sort_order?: number
  is_active?: boolean
  starts_at?: string | null
  ends_at?: string | null
}

export type StoreBannerUpdatePayload = {
  title_ar?: string | null
  title_en?: string | null
  image_url?: string | null
  target_url?: string | null
  sort_order?: number | null
  is_active?: boolean | null
  starts_at?: string | null
  ends_at?: string | null
}